# CS412 - Machine Learning - 2020

# Homework 2

100 pts

# Goal

The goal of this homework is to get familiar feature handling and cross validation.


# Dataset

German Credit Risk dataset, prepared by Prof. Hoffman, classifies each person as having a good or bad credit risk. The dataset that we use consists of both numerical and categorical features.



# Task

Build a k-NN classifier with scikit-learn library to classify people as bad or good risks for the german credit dataset. 

# Software

Documentation for the necessary functions can be accessed from the link below.

[http://scikit-learn.org/stable/supervised_learning.html](http://scikit-learn.org/stable/supervised_learning.html)

# Submission

Follow the instructions at the end.


# 1) Initialize

First, make a copy of this notebook in your drive

In [317]:
# Mount to your drive, in this way you can reach files that are in your drive
# Run this cell
# Go through the link that will be showed below
# Select your google drive account and copy authorization code and paste here in output and press enter
# You can also follow the steps from that link
# https://medium.com/ml-book/simplest-way-to-open-files-from-google-drive-in-google-colab-fae14810674 

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 2) Load Dataset

To start working for your homework, take a copy of the folder, given in the below link to your own google drive. You find the train and test data under this folder.

[https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing](https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing)

After copy the folder, copy the path of the train and test dataset to paste them in the below cell to load your data.


In [318]:
import pandas as pd

test_df = pd.read_csv('/content/drive/My Drive/german_credit_test.csv')
train_df = pd.read_csv('/content/drive/My Drive/german_credit_train.csv')

# 3) Optional - Analyze the Dataset 

You can use the functions of the pandas library to analyze your train dataset in detail - **this part is OPTIONAL - look around the data as you wish**.


*   Display the number of instances and features in the train ***(shape function can be used)**
*   Display 5 random examples from the train ***(sample function can be used)**
*   Display the information about each features ***(info method can be used)**



In [319]:
# Print shape
print("Train data dimensionality: ", train_df.shape)

# Print random 5 rows
print("Examples from train data:  ")
train_df.sample(5)


Train data dimensionality:  (800, 13)
Examples from train data:  


AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
367           A14        36           A34  ...               A143    A152    1
670           A11        20           A34  ...               A141    A151    2
722           A14        15           A32  ...               A141    A151    1
652           A14         6           A34  ...               A143    A152    1
393           A11        24           A31  ...               A143    A151    2

[5 rows x 13 columns]

In [320]:
#just to make sure that our test data has the same columns 
# Print shape
print("Test data dimensionality: ", test_df.shape)

# Print random 5 rows
print("Examples from test data:  ")
test_df.sample(5)

Test data dimensionality:  (200, 13)
Examples from test data:  


AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
199           A11        12           A34  ...               A143    A152    2
57            A12         6           A32  ...               A141    A152    1
176           A12        12           A32  ...               A143    A151    2
169           A11        24           A31  ...               A142    A152    2
1             A11        12           A34  ...               A143    A152    2

[5 rows x 13 columns]

In [321]:
# Print the information about the dataset
print("Information about train data ", )
print (train_df.info())
print("Information about test data ", )
print (test_df.info())


Information about train data 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   AccountStatus      800 non-null    object
 1   Duration           800 non-null    int64 
 2   CreditHistory      800 non-null    object
 3   CreditAmount       800 non-null    int64 
 4   SavingsAccount     800 non-null    object
 5   EmploymentSince    800 non-null    object
 6   PercentOfIncome    800 non-null    int64 
 7   PersonalStatus     800 non-null    object
 8   Property           800 non-null    object
 9   Age                800 non-null    int64 
 10  OtherInstallPlans  800 non-null    object
 11  Housing            720 non-null    object
 12  Risk               800 non-null    int64 
dtypes: int64(5), object(8)
memory usage: 81.4+ KB
None
Information about test data 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Dat

Hence we are now sure that our train and test data are of the same format

# 4) Define your train and test labels

*  Define labels for both train and test data in new arrays 
*  And remove the label column from both train and test sets do tht it is not used as a feature! 


(**you can use pop method**)


In [322]:
# Define labels


x_train = train_df.drop("Risk", axis=1)
y_train = train_df["Risk"]

x_test=test_df.drop("Risk", axis=1)
y_test=test_df["Risk"]



# 5) Handle missing values if any 

*   Print the columns that have **NaN** values (**isnull** method can be used)
*   You can impute missing values with mode of that feature or remove samples or attributes
*   To impute the test set, you should use the mode values that you obtain from **train** set, as **you should not be looking at your test data to gain any information or advantage.**



In [323]:
# Print columns with NaN values

print('columns that have NaN value in train set', x_train.columns[x_train.isnull().any()])
print('columns that have NaN value in test set', x_test.columns[x_test.isnull().any()])


columns that have NaN value in train set Index(['Housing'], dtype='object')
columns that have NaN value in test set Index(['Housing'], dtype='object')


In [324]:
# Impute missing values by replacing with mode value


x_train['Housing'] = x_train['Housing'].fillna(x_train['Housing'].mode()[0])

x_test['Housing'] = x_test['Housing'].fillna(x_train['Housing'].mode()[0])


In [325]:
x_train.isnull().sum()

AccountStatus        0
Duration             0
CreditHistory        0
CreditAmount         0
SavingsAccount       0
EmploymentSince      0
PercentOfIncome      0
PersonalStatus       0
Property             0
Age                  0
OtherInstallPlans    0
Housing              0
dtype: int64

# 6) Transform categorical / ordinal features

* Transform all categorical / ordinal features using the methods that you have learnt in lectures and recitation 4 for both train and test data
* You saw the dictionary use for mapping in recitation. (You can use **replace function** to assign new values to the categories of a column).

*  The class of the categorical attributes in the dataset are defined as follows:
  - Status of existing checking account
     - A11 :      ... <    0 DM
	- A12 : 0 <= ... <  200 DM
	- A13 :      ... >= 200 DM / salary assignments for at least 1 year
     - A14 : no checking account

 - Credit history
    - A30 : no credits taken/all credits paid back duly
    - A31 : all credits at this bank paid back duly
	- A32 : existing credits paid back duly till now
    - A33 : delay in paying off in the past
	- A34 : critical account/other credits existing (not at this bank)

  - Savings account
    - A61 :          ... <  100 DM
	- A62 :   100 <= ... <  500 DM
	- A63 :   500 <= ... < 1000 DM
	- A64 :          .. >= 1000 DM
    - A65 :   unknown/ no savings account

 - Employment Since
    - A71 : unemployed
    - A72 :       ... < 1 year
	- A73 : 1  <= ... < 4 years  
	- A74 : 4  <= ... < 7 years
	- A75 :       .. >= 7 years
 
 - Personal Status
    - A91 : male   : divorced/separated
	- A92 : female : divorced/separated/married
    - A93 : male   : single
	- A94 : male   : married/widowed
	- A95 : female : single

  - Property
     -  A121 : real estate
	- A122 : if not A121 : building society savings agreement/life insurance
    - A123 : if not A121/A122 : car or other, not in attribute 6
	- A124 : unknown / no property

 - OtherInstallPlans  
    - A141 : bank
	- A142 : stores
	- A143 : none

 - Housing
    -  A151 : rent
	 - A152 : own
	- A153 : for free

In [326]:
#mapping
import numpy as np

map_dict = {'A11':'... < 0 DM','A12':'0 <= ... < 200 DM','A13':'... >= 200 DM / salary assignments for at least 1 year','A14':'no checking account',
            'A30':'no credits taken/ all credits paid back duly','A31':'all credits at this bank paid back duly','A32':'existing credits paid back duly till now',
            'A33':'delay in paying off in the past','A34':'critical account/ other credits existing (not at this bank)',
            'A61':'... < 100 DM','A62':'100 <= ... < 500 DM','A63':'500 <= ... < 1000 DM','A64':'.. >= 1000 DM','A65':'no saving account','A71':'unemployed',
            'A72':'... < 1 year','A73':'1 <= ... < 4 years','A74':'4 <= ... < 7 years','A75':'.. >= 7 years','A91':'male - divorced/separated',
            'A92':'female - divorced/separated/married','A93':'male - single','A94':'male - married/widowed','A95':'female - single',
            'A121':'real estate','A122':'building society savings agreement/ life insurance',
            'A123':'car or other, not in attribute Account','A124':'no property','A141':'bank','A142':'stores','A143':'none','A151':'rent','A152':'own',
            'A153':'for free'}
good_credit_mapping = {2:0}


x_train = x_train.replace(map_dict)
y_train = y_train.replace(good_credit_mapping)

#same mapping for test data also
x_test= x_test.replace(map_dict)
y_test = y_test.replace(good_credit_mapping)

In [327]:
# Transform the categorical / ordinal attributes
x_train['AccountStatus'].unique()

array(['no checking account', '... < 0 DM', '0 <= ... < 200 DM',
       '... >= 200 DM / salary assignments for at least 1 year'],
      dtype=object)

In [328]:
status_map = {'no checking account':0,'... < 0 DM':1, '0 <= ... < 200 DM':2, '... >= 200 DM / salary assignments for at least 1 year':3}
x_train['AccountStatus'] = x_train['AccountStatus'].replace(status_map)
x_test['AccountStatus'] = x_test['AccountStatus'].replace(status_map)


In [329]:
x_train['SavingsAccount'].unique()

array(['no saving account', '... < 100 DM', '100 <= ... < 500 DM',
       '500 <= ... < 1000 DM', '.. >= 1000 DM'], dtype=object)

In [330]:
savings_map = {'no saving account':0, '... < 100 DM':1,'100 <= ... < 500 DM':2,'500 <= ... < 1000 DM':3,'.. >= 1000 DM':4}
x_train['SavingsAccount'] = x_train['SavingsAccount'].replace(savings_map)
x_test['SavingsAccount'] = x_test['SavingsAccount'].replace(savings_map)
x_train.head()

AccountStatus  Duration  ... OtherInstallPlans   Housing
0              0        12  ...              none       own
1              1         9  ...              none       own
2              1        18  ...              none  for free
3              1        14  ...              none       own
4              0        15  ...              none       own

[5 rows x 12 columns]

In [331]:
x_train['EmploymentSince'].unique()

array(['unemployed', '1 <= ... < 4 years', '.. >= 7 years',
       '4 <= ... < 7 years', '... < 1 year'], dtype=object)

In [332]:
employment_map = {'unemployed':0,'... < 1 year':1,'1 <= ... < 4 years':2,'4 <= ... < 7 years':3,'.. >= 7 years':4}
x_train['EmploymentSince'] = x_train['EmploymentSince'].replace(employment_map)
x_test['EmploymentSince'] = x_test['EmploymentSince'].replace(employment_map)

In [333]:
x_train.head()

AccountStatus  Duration  ... OtherInstallPlans   Housing
0              0        12  ...              none       own
1              1         9  ...              none       own
2              1        18  ...              none  for free
3              1        14  ...              none       own
4              0        15  ...              none       own

[5 rows x 12 columns]

In [334]:
x_train['CreditHistory'].unique()

array(['existing credits paid back duly till now',
       'critical account/ other credits existing (not at this bank)',
       'delay in paying off in the past',
       'all credits at this bank paid back duly',
       'no credits taken/ all credits paid back duly'], dtype=object)

In [335]:
creditHistory_map = {'delay in paying off in the past':0,
                  'critical account/ other credits existing (not at this bank)':1,
                  'existing credits paid back duly till now':2,
                  'no credits taken/ all credits paid back duly':3,
                  'all credits at this bank paid back duly':4}
x_train['CreditHistory'] = x_train['CreditHistory'].replace(creditHistory_map)
x_test['CreditHistory'] = x_test['CreditHistory'].replace(creditHistory_map)

In [336]:
x_train.head()

AccountStatus  Duration  CreditHistory  ...  Age  OtherInstallPlans   Housing
0              0        12              2  ...   38               none       own
1              1         9              2  ...   25               none       own
2              1        18              1  ...   36               none  for free
3              1        14              2  ...   45               none       own
4              0        15              2  ...   40               none       own

[5 rows x 12 columns]

In [337]:
x_train['Housing'].unique()

array(['own', 'for free', 'rent'], dtype=object)

In [338]:
housing_map = {'rent':0,
                  'for free':1,
                  'own':2}
x_train['Housing'] = x_train['Housing'].replace(housing_map)
x_test['Housing'] = x_test['Housing'].replace(housing_map)

In [339]:
x_train.head()

AccountStatus  Duration  CreditHistory  ...  Age  OtherInstallPlans  Housing
0              0        12              2  ...   38               none        2
1              1         9              2  ...   25               none        2
2              1        18              1  ...   36               none        1
3              1        14              2  ...   45               none        2
4              0        15              2  ...   40               none        2

[5 rows x 12 columns]

In [ ]:
from sklearn.preprocessing import OneHotEncoder

enc = OneHotEncoder(handle_unknown='ignore')
enc.fit_transform(x_train[['PersonalStatus']]).toarray()


enc_test = OneHotEncoder(handle_unknown='ignore')
enc_test.fit_transform(x_test[['PersonalStatus']]).toarray()


In [341]:
enc.categories_

[array(['female - divorced/separated/married', 'male - divorced/separated',
        'male - married/widowed', 'male - single'], dtype=object)]

In [342]:
dummy_personalStatus = enc.fit_transform(x_train[['PersonalStatus']]).toarray()
dummy_personalStatus = pd.DataFrame(dummy_personalStatus,columns=enc.categories_)
dummy_personalStatus.head()

dummy_personalStatus_test = enc_test.fit_transform(x_test[['PersonalStatus']]).toarray()
dummy_personalStatus_test = pd.DataFrame(dummy_personalStatus_test,columns=enc_test.categories_)
dummy_personalStatus_test.head()

female - divorced/separated/married  ... male - single
0                                 0.0  ...           1.0
1                                 1.0  ...           0.0
2                                 0.0  ...           0.0
3                                 0.0  ...           1.0
4                                 1.0  ...           0.0

[5 rows x 4 columns]

In [343]:
x_train = pd.merge(x_train,dummy_personalStatus,right_index=True,left_index=True)
x_train = x_train.drop(columns='PersonalStatus')
x_test = pd.merge(x_test,dummy_personalStatus_test,right_index=True,left_index=True)
x_test = x_test.drop(columns='PersonalStatus')
x_train.head()




AccountStatus  Duration  ...  (male - married/widowed,)  (male - single,)
0              0        12  ...                        0.0               1.0
1              1         9  ...                        0.0               1.0
2              1        18  ...                        0.0               1.0
3              1        14  ...                        0.0               0.0
4              0        15  ...                        0.0               1.0

[5 rows x 15 columns]

In [ ]:
enc = OneHotEncoder(handle_unknown='ignore')
enc.fit_transform(x_train[['Property']]).toarray()


enc_test = OneHotEncoder(handle_unknown='ignore')
enc_test.fit_transform(x_test[['Property']]).toarray()

In [345]:
enc.categories_

[array(['building society savings agreement/ life insurance',
        'car or other, not in attribute Account', 'no property',
        'real estate'], dtype=object)]

In [346]:
dummy_property = enc.fit_transform(x_train[['Property']]).toarray()
dummy_property = pd.DataFrame(dummy_property,columns=enc.categories_)

dummy_property_test = enc_test.fit_transform(x_test[['Property']]).toarray()
dummy_property_test = pd.DataFrame(dummy_property_test,columns=enc_test.categories_)


In [347]:
x_train = pd.merge(x_train,dummy_property,right_index=True,left_index=True)
x_train = x_train.drop(columns='Property')
x_test = pd.merge(x_test,dummy_property_test,right_index=True,left_index=True)
x_test = x_test.drop(columns='Property')
x_train.head()

AccountStatus  Duration  ...  (no property,)  (real estate,)
0              0        12  ...             1.0             0.0
1              1         9  ...             0.0             1.0
2              1        18  ...             1.0             0.0
3              1        14  ...             0.0             0.0
4              0        15  ...             0.0             0.0

[5 rows x 18 columns]

In [ ]:
enc = OneHotEncoder(handle_unknown='ignore')
enc.fit_transform(x_train[['OtherInstallPlans']]).toarray()


enc_test = OneHotEncoder(handle_unknown='ignore')
enc_test.fit_transform(x_test[['OtherInstallPlans']]).toarray()

In [349]:
enc.categories_

[array(['bank', 'none', 'stores'], dtype=object)]

In [350]:
dummy_OtherInstallPlans = enc.fit_transform(x_train[['OtherInstallPlans']]).toarray()
dummy_OtherInstallPlans = pd.DataFrame(dummy_OtherInstallPlans,columns=enc.categories_)


dummy_OtherInstallPlans_test = enc_test.fit_transform(x_test[['OtherInstallPlans']]).toarray()
dummy_OtherInstallPlans_test = pd.DataFrame(dummy_OtherInstallPlans_test,columns=enc_test.categories_)
dummy_OtherInstallPlans_test.head()

bank none stores
0  0.0  0.0    1.0
1  0.0  1.0    0.0
2  0.0  1.0    0.0
3  0.0  1.0    0.0
4  0.0  1.0    0.0

In [351]:
x_train = pd.merge(x_train,dummy_OtherInstallPlans,right_index=True,left_index=True)
x_train = x_train.drop(columns='OtherInstallPlans')
x_test = pd.merge(x_test,dummy_OtherInstallPlans_test,right_index=True,left_index=True)
x_test = x_test.drop(columns='OtherInstallPlans')
x_train.head()

AccountStatus  Duration  CreditHistory  ...  (bank,)  (none,)  (stores,)
0              0        12              2  ...      0.0      1.0        0.0
1              1         9              2  ...      0.0      1.0        0.0
2              1        18              1  ...      0.0      1.0        0.0
3              1        14              2  ...      0.0      1.0        0.0
4              0        15              2  ...      0.0      1.0        0.0

[5 rows x 20 columns]

# 7) Build a k-NN classifier on training data and perform models selection using 5 fold cross validation

*  Initialize k-NN classifiers with **k= 5, 10, 15**
*  Calculate the cross validation scores using cross_al_score method, number of folds is 5. 
*  Note: Xval is performed on training data! Do not use test data in any way and do not separate a hold-out validation set, rather use cross-validation.

Documentation of the cross_val_score method:

[https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score)

*  Stores the average accuracies of these folds
*  Select the value of k using the cross validation results. 

In [352]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from statistics import mean

# k values
kVals = [5,10,15]

# Save the accuracies of each value of kVal in [accuracies] variable
accuracies = []

# Loop over values of k for the k-Nearest Neighbor classifier
for k in kVals:
  # Initialize a k-NN classifier with k neighbors
  knn = KNeighborsClassifier(n_neighbors=k)


  # Calculate the 5 fold cross validation scores using cross_val_score
  # cv parameter: number of folds, in our case it must be 5
  acc =cross_val_score(knn,x_train,y_train, cv = 5)
  scores = acc.mean()
  accuracies.append(scores)

  # Stores the average accuracies of the scores in accuracies variable, you can use mean method


print(accuracies)

[0.6762499999999999, 0.67875, 0.7100000000000001]


# 8) Retrain using all training data and test on test set

* Train a classifier with the chosen k value of the best classifier using **all training data**. 

Note:  k-NN training involves no explicit training, but this is what we would do after model selection with decision trees or any other ML approach (we had 5 diff. models -one for each fold - for each k in the previous step - dont know which one to submit. Even if we picked the best one, it does not use all training samples.

* Predict the labels of testing data 

* Report the accuracy 

In [353]:
from sklearn.metrics import accuracy_score

# Train the best classifier using all training set
best_knn = KNeighborsClassifier(n_neighbors=15)
best_knn.fit(x_train,y_train)

# Estimate the prediction of the test data
y_predTest=best_knn.predict(x_test)

# Print accuracy of test data
print(accuracy_score(y_test, y_predTest))


0.665


# 9) Bonus (5pts)

There is a limited bonus for any extra work that you may use and improve the above results. 

You may try a larger k values, scale input features, remove some features, .... Please **do not overdo**, maybe spend another 30-60min on this. The idea is not do an exhaustive search (which wont help your understanding of ML process), but just to give some extra points to those who may look at the problem a little more comprehensively. 

**If you obtain better results than the above, please indicate the best model you have found and the corresponding accuracy.**

E.g. using feature normalization ..... and removing .... features and using a value k=...., I have obtained ....% accuracy.


In [354]:
x_train_bonus=x_train.drop("Age", axis=1)
x_test_bonus=x_test.drop("Age", axis=1)


In [355]:
kVals = [10,12,13,14,15,16]
for k in kVals:
  knn = KNeighborsClassifier(n_neighbors=k)
  knn.fit(x_train_bonus,y_train)
  y_pred_knn = knn.predict(x_test_bonus)
  print('for k=',k,'accuracy score:', accuracy_score(y_test,y_pred_knn))

for k= 10 accuracy score: 0.63
for k= 12 accuracy score: 0.64
for k= 13 accuracy score: 0.67
for k= 14 accuracy score: 0.66
for k= 15 accuracy score: 0.67
for k= 16 accuracy score: 0.635


In [356]:
x_train_bonus=x_train.drop("CreditAmount", axis=1)


x_test_bonus=x_test.drop("CreditAmount", axis=1)

In [357]:
kVals = [10,12,13,14,15,16,20,25,30,7,8]
for k in kVals:
  knn = KNeighborsClassifier(n_neighbors=k)
  knn.fit(x_train_bonus,y_train)
  y_pred_knn = knn.predict(x_test_bonus)
  print('for k=',k,'accuracy score:', accuracy_score(y_test,y_pred_knn))

for k= 10 accuracy score: 0.645
for k= 12 accuracy score: 0.67
for k= 13 accuracy score: 0.685
for k= 14 accuracy score: 0.675
for k= 15 accuracy score: 0.68
for k= 16 accuracy score: 0.685
for k= 20 accuracy score: 0.66
for k= 25 accuracy score: 0.69
for k= 30 accuracy score: 0.68
for k= 7 accuracy score: 0.66
for k= 8 accuracy score: 0.635


By removing Age and CreditAmount features and using a value k=25, I have obtained 69% accuracy.

# 10) Notebook & Report


This homeworsk aims at building a k-NN classifier using Scikit library, to classify people as bad or good for credit risks. We used German Credit Risk dataset, consisting of both numerical and ordinal/categorical features. Our problem necessites the transfering and mapping categorical/ordinal features and choosing k value in such a way that we obtain highest accuracy. When we download the data, we first define the label for both train and test set and, by dropping the label column which is Risk, from datasets we hence obtain x_train/test(features df) and y_train/test(label) After that we check if there is any missing value with isnull() function. We see that there are NaN values in housing feature hence we impute those ones for both train and set set with the mode value of housing of train data. Notice that we do not use mode of test data to avoid biased results. Before training our knn model, we must transform all categorical/ordinal features(basically non-numerical values)into numbers because kNN cannot work with them directly, unlike decision trees. We use dictionary to map the corresponding meanings of categories in the dataset for example A14 means no checking account. After mapping, We first decide which ones are the ordinal variables,it is when there is a ordering of the categories. To determine those, we need some domain knowledge.Some categories are bad for the credit such as no credit account will be the worst case hence we show it as 0 and we show other ones as 1,2 depending on the importance and 'betterness'. Better it is, higher the integer value. In this homework, Account Status, Savings Account and Employement Since, Housing and Credit History features are considered as ordinal. Note that this choice may be different, it depends on the person who writes the code but result may change accordingly. Categorical variables don't have an ordering to categories. In this homework Personal Status, Property and Other Install Plans are considered as categorical. For the  categorical data we use one hot encoder, it converts each category value into a new column and assigned a 1 or 0. This approach prevents hierarchy. Finally, we build kNN classifers with k=5,10,15 and we calculate 5 fold cross validation scores. 

| AVERAGE VALIDATION ACCURACIES FOR 5 FOLD CV         |        |  
|---------------------|--------|
| kNN model with k=5  | 0.6762 |  
| kNN model with k=10 | 0.6787 |  
| kNN model with k=15 | 0.71   |   

We have obtained the best result with the kNN classifier with k=15, so we train the kNN classifier with the k=15 using all training data. After, we apply this model to test data and we obtain an accuracy of 66.5% 
To further develop this model, we drop Age and Credit Amount features from test and train data because they seem irrelevant for the problem. By doing so, we obtain 69% accuracy with kNN classifier where k=25.